### PR TITLE
Add PlayerShieldDisableEvent

### PR DESCRIPTION
**Pull request by Cryptite**

### CODE REVIEW

File: patches/api/0333-Add-PlayerShieldDisableEvent.patch [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+Subject: [PATCH] Add PlayerShieldDisableEvent
+
+
+diff --git a/src/main/java/org/bukkit/event/player/PlayerShieldDisableEvent.java b/src/main/java/org/bukkit/event/player/PlayerShieldDisableEvent.java
+new file mode 100644
+index 0000000000000000000000000000000000000000..93f55f13c0fdd100dc87b2acf0127eb518205f29
+--- /dev/null
++++ b/src/main/java/org/bukkit/event/player/PlayerShieldDisableEvent.java
+@@ -0,0 +1,44 @@
++package org.bukkit.event.player;
```
Comment (lynxplay) [inline patches/api/0333-Add-PlayerShieldDisableEvent.patch]: This should probably be in the `io.papermc.paper.event.player` package.

File: patches/api/0333-Add-PlayerShieldDisableEvent.patch [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
++
++import org.bukkit.entity.Entity;
++import org.bukkit.entity.Player;
++import org.bukkit.event.Event;
++import org.bukkit.event.HandlerList;
++import org.jetbrains.annotations.NotNull;
++import org.jetbrains.annotations.Nullable;
++
++/**
++ * Thrown whenever a {@link Player} dies
```
Comment (lynxplay) [inline patches/api/0333-Add-PlayerShieldDisableEvent.patch]: This is definitely not the correct java documentation. Please update it to reflect this event.

File: patches/api/0333-Add-PlayerShieldDisableEvent.patch [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
++ */
++public class PlayerShieldDisableEvent extends Event {
```
Comment (lynxplay) [inline patches/api/0333-Add-PlayerShieldDisableEvent.patch]: Any reason this is not extending the [PlayerEvent](https://papermc.io/javadocs/paper/1.17/org/bukkit/event/player/PlayerEvent.html) class ?

File: patches/server/0807-Add-PlayerShieldDisableEvent.patch [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+ import net.minecraft.world.level.storage.loot.LootContext;
++import org.bukkit.Bukkit; // Paper
+ import org.bukkit.craftbukkit.event.CraftEventFactory;
+ import org.bukkit.craftbukkit.entity.CraftLivingEntity;
+ import org.bukkit.event.entity.CreatureSpawnEvent;
+@@ -81,6 +82,7 @@ import org.bukkit.event.entity.EntityTargetEvent;
+ import org.bukkit.event.entity.EntityTransformEvent;
+ import org.bukkit.event.entity.EntityUnleashEvent;
+ import org.bukkit.event.entity.EntityUnleashEvent.UnleashReason;
++import org.bukkit.event.player.PlayerShieldDisableEvent; // Paper
```
Comment (lynxplay) [inline patches/server/0807-Add-PlayerShieldDisableEvent.patch]: This import should be ~inlined to reduce the diff~ replaced by a fully qualified name.
The previous Bukkit import is useless as you can replace it with 
```java
event.callEvent();
```

File: patches/server/0807-Add-PlayerShieldDisableEvent.patch [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+ 
+ public abstract class Mob extends LivingEntity {
+@@ -1590,6 +1592,11 @@ public abstract class Mob extends LivingEntity {
+             if (this.random.nextFloat() < f) {
+                 player.getCooldowns().addCooldown(Items.SHIELD, 100);
+                 this.level.broadcastEntityEvent(player, (byte) 30);
++
++                // Paper start
++                PlayerShieldDisableEvent event = new PlayerShieldDisableEvent((org.bukkit.entity.Player) player.getBukkitEntity(), getBukkitEntity());
++                Bukkit.getPluginManager().callEvent(event);
```
Comment (lynxplay) [inline patches/server/0807-Add-PlayerShieldDisableEvent.patch]: See previous comment on import as to why this line should be replaced.

File: patches/server/0807-Add-PlayerShieldDisableEvent.patch [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+ // CraftBukkit start
++import org.bukkit.Bukkit; // Paper
+ import org.bukkit.craftbukkit.entity.CraftHumanEntity;
+ import org.bukkit.craftbukkit.event.CraftEventFactory;
+ import org.bukkit.craftbukkit.util.CraftVector;
+@@ -123,6 +124,7 @@ import org.bukkit.event.entity.CreatureSpawnEvent;
+ import org.bukkit.event.entity.EntityCombustByEntityEvent;
+ import org.bukkit.event.entity.EntityExhaustionEvent;
+ import org.bukkit.event.player.PlayerDropItemEvent;
++import org.bukkit.event.player.PlayerShieldDisableEvent; // Paper
```
Comment (lynxplay) [inline patches/server/0807-Add-PlayerShieldDisableEvent.patch]: See previous import comments.

File: patches/server/0807-Add-PlayerShieldDisableEvent.patch [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+             this.getCooldowns().addCooldown(Items.SHIELD, 100);
+             this.stopUsingItem();
+             this.level.broadcastEntityEvent(this, (byte) 30);
+-        }
+ 
++            PlayerShieldDisableEvent event = new PlayerShieldDisableEvent((org.bukkit.entity.Player) getBukkitEntity(), attacker.getBukkitEntity());
++            Bukkit.getPluginManager().callEvent(event);
++        }
+     }
++    // Paper end
```
Comment (lynxplay) [inline patches/server/0807-Add-PlayerShieldDisableEvent.patch:73]: This diff seems a bit too big. See [here](https://github.com/PaperMC/Paper/blob/master/patches/api/0165-Support-cancellation-supression-of-EntityDismount-Ve.patch#L37) for an example on how to create an overloaded method using paper comments.

**Review by lynxplay — CHANGES_REQUESTED** [review]
Besides the specifics, is there any reason this event cannot implement [Cancellable](https://papermc.io/javadocs/paper/1.17/org/bukkit/event/Cancellable.html) ? Seems rather straight forward.

**Comment by Cryptite** [comment]
Ruh roh, something bad happened with rebuilding, inc another commit methinks

**Pull request closed without merging**